File: src/creational/factory-method/FactoryMethod.cs
namespace DesignPatterns.Creational.FactoryMethod
{
    // Product interface
    public interface IIceCream
    {
        void Serve();
    }

    // Concrete Product: Cone
    public class ConeIceCream : IIceCream
    {
        public void Serve()
        {
            Console.WriteLine("Serving ice cream in a cone.");
        }
    }

    // Concrete Product: Cup
    public class CupIceCream : IIceCream
    {
        public void Serve()
        {
            Console.WriteLine("Serving ice cream in a cup.");
        }
    }

    // Creator abstract class
    public abstract class IceCreamCreator
    {
        public abstract IIceCream FactoryMethod();

        public void ServeIceCream()
        {
            var iceCream = FactoryMethod();
            iceCream.Serve();
        }
    }

    // Concrete Creator: Cone
    public class ConeIceCreamCreator : IceCreamCreator
    {
        public override IIceCream FactoryMethod()
        {
            return new ConeIceCream();
        }
    }

    // Concrete Creator: Cup
    public class CupIceCreamCreator : IceCreamCreator
    {
        public override IIceCream FactoryMethod()
        {
            return new CupIceCream();
        }
    }

    // Client code
    class Program
    {
        static void Main()
        {
            Console.WriteLine("Choose your ice cream type (cone/cup):");
            string? input = Console.ReadLine()?.ToLower();

            IceCreamCreator creator = input switch
            {
                "cone" => new ConeIceCreamCreator(),
                "cup" => new CupIceCreamCreator(),
                _ => throw new ArgumentException("Invalid type")
            };

            creator.ServeIceCream();
        }
    }
}

File: src/creational/factory-method/factory_method.py
# Product interface
class IceCream:
    def serve(self):
        pass

# Concrete Product: Cone
class ConeIceCream(IceCream):
    def serve(self):
        print("Serving ice cream in a cone.")

# Concrete Product: Cup
class CupIceCream(IceCream):
    def serve(self):
        print("Serving ice cream in a cup.")

# Creator abstract class
class IceCreamCreator:
    def factory_method(self):
        pass

    def serve_ice_cream(self):
        ice_cream = self.factory_method()
        ice_cream.serve()

# Concrete Creator: Cone
class ConeIceCreamCreator(IceCreamCreator):
    def factory_method(self):
        return ConeIceCream()

# Concrete Creator: Cup
class CupIceCreamCreator(IceCreamCreator):
    def factory_method(self):
        return CupIceCream()

# Client code
if __name__ == "__main__":
    ice_cream_type = input("Choose your ice cream type (cone/cup): ").lower()

    if ice_cream_type == "cone":
        creator = ConeIceCreamCreator()
    elif ice_cream_type == "cup":
        creator = CupIceCreamCreator()
    else:
        raise ValueError("Invalid ice cream type.")

    creator.serve_ice_cream()
